Constraints on the max width/height cause blurry getDisplayMedia video on Safari 17

youennf opened the pull request:
#### 2a1363e51a7cb864069914e8c28bfc1aae1e6d06
<pre>
Constraints on the max width/height cause blurry getDisplayMedia video on Safari 17
<a href="https://bugs.webkit.org/show_bug.cgi?id=263015">https://bugs.webkit.org/show_bug.cgi?id=263015</a>
rdar://116810370

Reviewed by Eric Carlson.

We were applying constraints at the time of starting the capture.
For getDisplayMedia, we do not know yet the size of window/screen, so we start with 640x480.
Later on, the size may increase, but we were not recomputing the constraints and our resizer was sticking to the old values.

Whenever there is a change of size, we are now calling configuration change.
Within UserMediaCaptureManagerProxy, in case of getDisplayMedia, we will reapply constraints.
This will trigger recomputation of the resizer based on any max constraint.

We add a test that is emulating the change of screen being captured.
We update an existing testRunner API for that purpose.

* LayoutTests/fast/mediastream/getDisplayMedia-max-constraints4-expected.txt: Added.
* LayoutTests/fast/mediastream/getDisplayMedia-max-constraints4.html: Added.
* LayoutTests/fast/mediastream/mediastreamtrack-configurationchange.html:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.cpp:
(WebCore::DisplayCaptureSourceCocoa::emitFrame):
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
(WebCore::ScreenCaptureKitCaptureSource::streamDidOutputVideoSampleBuffer):
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::MockDisplayCapturer::triggerMockCaptureConfigurationChange):
(WebCore::MockRealtimeDisplaySourceFactory::latestCapturer):
(WebCore::MockRealtimeMediaSourceCenter::triggerMockCaptureConfigurationChange):
(WebCore::MockRealtimeMediaSourceCenter::triggerMockMicrophoneConfigurationChange): Deleted.
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.h:
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::triggerMockCaptureConfigurationChange):
(WebKit::GPUProcess::triggerMockMicrophoneConfigurationChange): Deleted.
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageTriggerMockCaptureConfigurationChange):
(WKPageTriggerMockMicrophoneConfigurationChange): Deleted.
* Source/WebKit/UIProcess/API/C/WKPagePrivate.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::updateVideoConstraints):
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::applyConstraints):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::triggerMockCaptureConfigurationChange):
(WebKit::GPUProcessProxy::triggerMockMicrophoneConfigurationChange): Deleted.
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::triggerMockCaptureConfigurationChange):
(WTR::TestRunner::triggerMockMicrophoneConfigurationChange): Deleted.
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::triggerMockCaptureConfigurationChange):
(WTR::TestController::triggerMockMicrophoneConfigurationChange): Deleted.
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):

Canonical link: <a href="https://commits.webkit.org/269406@main">https://commits.webkit.org/269406@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4ed88cce4f4d33e3e449a0ed33629fe6381df6d9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22480 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/119 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23558 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24384 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20808 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22735 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/953 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23011 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21792 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22719 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/95 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19512 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25236 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/71 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20369 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26610 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/20492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20610 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24466 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/72 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17919 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/54 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20177 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5352 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/85 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/82 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson reviewed (approved):
r=me once the bots are happy

youennf committed:
Constraints on the max width/height cause blurry getDisplayMedia video on Safari 17
https://bugs.webkit.org/show_bug.cgi?id=263015
rdar://116810370

Reviewed by Eric Carlson.

We were applying constraints at the time of starting the capture.
For getDisplayMedia, we do not know yet the size of window/screen, so we start with 640x480.
Later on, the size may increase, but we were not recomputing the constraints and our resizer was sticking to the old values.

Whenever there is a change of size, we are now calling configuration change.
Within UserMediaCaptureManagerProxy, in case of getDisplayMedia, we will reapply constraints.
This will trigger recomputation of the resizer based on any max constraint.

We add a test that is emulating the change of screen being captured.
We update an existing testRunner API for that purpose.

* LayoutTests/fast/mediastream/getDisplayMedia-max-constraints4-expected.txt: Added.
* LayoutTests/fast/mediastream/getDisplayMedia-max-constraints4.html: Added.
* LayoutTests/fast/mediastream/mediastreamtrack-configurationchange.html:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.cpp:
(WebCore::DisplayCaptureSourceCocoa::emitFrame):
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm:
(WebCore::ScreenCaptureKitCaptureSource::streamDidOutputVideoSampleBuffer):
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::MockDisplayCapturer::triggerMockCaptureConfigurationChange):
(WebCore::MockRealtimeDisplaySourceFactory::latestCapturer):
(WebCore::MockRealtimeMediaSourceCenter::triggerMockCaptureConfigurationChange):
(WebCore::MockRealtimeMediaSourceCenter::triggerMockMicrophoneConfigurationChange): Deleted.
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.h:
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::triggerMockCaptureConfigurationChange):
(WebKit::GPUProcess::triggerMockMicrophoneConfigurationChange): Deleted.
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageTriggerMockCaptureConfigurationChange):
(WKPageTriggerMockMicrophoneConfigurationChange): Deleted.
* Source/WebKit/UIProcess/API/C/WKPagePrivate.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::updateVideoConstraints):
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::applyConstraints):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::triggerMockCaptureConfigurationChange):
(WebKit::GPUProcessProxy::triggerMockMicrophoneConfigurationChange): Deleted.
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::triggerMockCaptureConfigurationChange):
(WTR::TestRunner::triggerMockMicrophoneConfigurationChange): Deleted.
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::triggerMockCaptureConfigurationChange):
(WTR::TestController::triggerMockMicrophoneConfigurationChange): Deleted.
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):

Canonical link: https://commits.webkit.org/269406@main
